### PR TITLE
fix(sandbox-runtime): scope post-compaction message acceptance to the active prompt

### PR DESCRIPTION
**Pull request by ColeMurray**
## Summary

After a context compaction, the session timeline could end a turn by displaying a **repeat of an earlier turn's final message** instead of the turn's real closing message. The timeline rendered exactly what was persisted — the corruption happened in the bridge.

## Root cause

When `session.compacted` fires, the bridge sets `compaction_occurred` and falls back to accepting assistant messages whose `parentID` no longer matches (compaction rewrites the message chain). That fallback was unscoped: it matched **every** non-summary assistant message in the OpenCode session, including messages from **previous turns**.

The visible damage comes from `_fetch_final_message_state`, which runs at idle (and on the timeout/disconnect paths) against `GET /session/:id/message` — the **full session history**. Prior turns' text parts were never streamed during the current prompt, so every one of them reads as "longer than sent" (`previously_sent = ""`) and gets re-emitted as a `token` event bound to the **current** control-plane message ID. The control plane upserts tokens into one row per segment, so the last re-emitted part — an old turn's text — overwrites the current turn's final output right before `execution_complete`.

Observed in production: a compaction turn ended by displaying a verbatim copy of an earlier turn's review report (timestamped at completion), while the turn's real final message was never shown. The same over-broad predicate also existed on the live SSE path (`_on_message_updated`), where it could authorize any old assistant message OpenCode re-touches after compaction.

## Fix

Scope the compaction fallback to messages created during this prompt. OpenCode message IDs are ascending (12 hex chars of `timestamp_ms * 0x1000 + counter`, fixed length — lexicographic order is creation order), and the bridge already generates its user-message ID with the same scheme and relies on this ordering (`stream_prompt` docstring). The new `_compaction_fallback_accepts` requires:

```
compaction_occurred and oc_msg_id > state.opencode_message_id
```

Applied in both `_fetch_final_message_state` and the `_on_message_updated` acceptance/error paths. Post-compaction continuations (created after the prompt's user message) still pass; prior turns' messages never do.

## Notes

- Verified against OpenCode v1.18.11 source: `Session.messages` returns full history (newest-first for ≤50 messages), the compaction summary is `summary: true` with `parentID` = the compaction user message, and `Identifier.create` makes ascending IDs lexicographically time-ordered.
- Turns already affected in stored timelines keep their overwritten text; the real final text was never persisted by the control plane.

## Verification

- `uv run --extra dev pytest tests/` — 723 passed
- New tests fail against the unfixed code (verified via stash) and pass with the fix:
  - `test_post_compaction_prior_prompt_message_is_not_accepted`
  - `test_post_compaction_error_on_prior_prompt_message_is_ignored`
  - `test_post_compaction_later_message_is_accepted`
  - `test_compaction_fallback_skips_prior_prompt_messages` (models the incident: history replay must not clobber the continuation's backfill)
- `uv run --extra dev ruff check` / `ruff format --check` — clean
- `uv run --extra dev mypy src/sandbox_runtime/prompt_stream.py` — clean

## Summary by CodeRabbit

* **Bug Fixes**
  * Prevented previous-session assistant responses from being replayed after conversation compaction.
  * Ensured only messages created after the current prompt are accepted during recovery and streaming.
  * Improved handling of older-message errors so they do not interrupt current output.
  * Ensured continuation responses are streamed correctly, including messages sharing the same timestamp.

* **Tests**
  * Added coverage for post-compaction message filtering, continuation output, message ordering, and final-message recovery.

### CODE REVIEW

File: packages/sandbox-runtime/src/sandbox_runtime/prompt_stream.py [hunk 1]
```diff
@@ -405,7 +405,10 @@ def _on_message_updated(
                 belongs_to_prompt = (
                     parent_matches
                     or oc_msg_id in state.correlated_compaction_summary_ids
-                    or (state.compaction_occurred and not is_compaction_summary)
+                    or (
+                        not is_compaction_summary
+                        and self._compaction_fallback_accepts(state, oc_msg_id)
+                    )
```
[the remark this answers is not in the document]
Comment (ColeMurray) [reply packages/sandbox-runtime/src/sandbox_runtime/prompt_stream.py:411]: Verified against OpenCode v1.18.11 source before scoping the fix this way — the described replay path has no trigger today:

- The `/event` SSE endpoint registers a live listener at connect and emits only `server.connected` plus subsequent events (`server/routes/instance/httpapi/handlers/event.ts`) — prior turns' `message.updated` events are never replayed into a fresh prompt stream.
- Compaction creates only new messages (summary + continue user message, `session/compaction.ts`); it does not republish prior user messages mid-prompt.
- `_fetch_final_message_state` skips `role != assistant` rows entirely, and its assistant acceptance now carries the ID boundary.

Gating `parent_matches` (or the user-ID discovery path) by ID order isn't free, either: the bridge's proposed user-message ID and OpenCode's server-side IDs come from independent per-process counters, so a same-millisecond tie can order the prompt's own chain ambiguously — an ordering gate there can false-reject legitimate current-prompt messages, which is a worse failure mode than the currently-unreachable replay. The fallback branch is different: the alternative there was accepting everything, which is the demonstrated prod bug this PR fixes.

A stricter continuation-chain model is a fair ask — tracking it in #1388 (extracting a message-attribution object) rather than growing the predicates in place here.

File: packages/sandbox-runtime/src/sandbox_runtime/prompt_stream.py [hunk 2]
```diff
@@ -450,6 +456,21 @@ def _on_message_updated(
 
         return []
 
+    @staticmethod
+    def _compaction_fallback_accepts(state: _PromptState, oc_msg_id: str) -> bool:
```
[the remark this answers is not in the document]
Comment (ColeMurray) [reply packages/sandbox-runtime/src/sandbox_runtime/prompt_stream.py:460]: Agreed on the direction — the eligibility policy is now stated in three places and this file shouldn't keep accreting attribution branches. I kept this PR to the minimal behavioral fix for the prod incident (post-compaction turns displaying a prior turn's final message), with the shared predicate extracted so the formula at least exists once.

Opened #1388 for the decomposition: a correlation object owning parent matching, summary handling, tracked IDs, and the post-prompt ID boundary, exposing one accept/reject decision to both the live SSE path and final-state reconciliation.

File: packages/sandbox-runtime/tests/test_prompt_stream.py [hunk 3]
```diff
@@ -644,6 +644,115 @@ def test_child_compaction_does_not_emit_parent_marker(self):
...
+        stream._apply_sse_event(
+            state,
+            sse(
+                "message.part.updated",
+                {
+                    "part": {
+                        "type": "text",
+                        "id": "part-prior",
+                        "sessionID": PARENT_SESSION_ID,
+                        "messageID": "msg_a-prior-turn",
```
[the remark this answers is not in the document]
Comment (ColeMurray) [reply packages/sandbox-runtime/tests/test_prompt_stream.py]: Done in bc87e3ca3. Added `oc_message_id(timestamp_ms, counter, suffix)` to `tests/conftest.py` mirroring the real format (12-hex time+counter segment + 14 base62 chars), plus a format-lock test asserting it reproduces `OpenCodeIdentifier.ascending`'s encoding, so the fixtures can't drift from the production ID contract.

The boundary tests now use valid-format IDs placed deterministically around the prompt's user message (60s before, same-ms, and after), including a same-timestamp counter-boundary case: `(T, counter-1)` rejected, `(T, counter+1)` accepted around the prompt's `(T, counter)`. The reconciliation regression test uses real-format IDs for the full chain (prior turn, compaction user, summary, synthetic continue, continuation) as well.
